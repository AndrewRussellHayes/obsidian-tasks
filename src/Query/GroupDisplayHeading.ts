/**
 * GroupDisplayHeading contains the data needed to render one heading for a group of tasks
 */
export class GroupDisplayHeading {
    /**
     * How nested the heading is.
     * 0 is the first group, meaning this heading was generated by
     * the first 'group by' instruction.
     */
    public readonly nestingLevel: number;

    /**
     * The text to be displayed for the group.
     */
    public readonly displayName: string;

    /**
     * Construct a GroupDisplayHeading object
     * @param {number} nestingLevel - See this.nestingLevel for details
     * @param {string} displayName - The text to be displayed for the group
     */
    constructor(nestingLevel: number, displayName: string) {
        this.nestingLevel = nestingLevel;
        this.displayName = displayName;
    }
}
