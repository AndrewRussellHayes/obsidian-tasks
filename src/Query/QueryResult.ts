import type { TaskGroups } from './TaskGroups';

export class QueryResult {
    public groups: TaskGroups;

    constructor(groups: TaskGroups) {
        this.groups = groups;
    }
}
