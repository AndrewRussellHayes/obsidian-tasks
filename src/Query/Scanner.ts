function endsWith1Slash(inputLine: string) {
    return inputLine.endsWith('\\');
}

function stripLeadingWhitespace(adjustedInputLine: string) {
    return adjustedInputLine.replace(/^[ \t]*/, '');
}

function stripEndingSlashAndPrecedingWhitespace(adjustedInputLine: string) {
    return adjustedInputLine.replace(/[ \t]*\\$/, '');
}

function adjustLine(inputLine: string, joinToNext: boolean) {
    let adjustedLine = inputLine;
    if (joinToNext) {
        // The new line will be appended to the previous one,
        // so discard any leading white space:
        adjustedLine = stripLeadingWhitespace(inputLine);
    }
    if (endsWith1Slash(inputLine)) {
        // This is a continuation line, so remove its trailing backslash
        // and any spare white space beforehand.
        adjustedLine = stripEndingSlashAndPrecedingWhitespace(adjustedLine);
    }
    return adjustedLine;
}

function saveLine(outputLines: string[], joinToNext: boolean, adjustedLine: string) {
    if (joinToNext) {
        outputLines[outputLines.length - 1] += ' ' + adjustedLine;
    } else {
        outputLines.push(adjustedLine);
    }
}

/**
 * Removes newlines escaped by a backslash.
 * A trailing backslash at the end of a line can be escaped by doubling it.
 *
 * @param input input string
 * @returns modified input
 */
export function continue_lines(input: string): string {
    const outputLines: string[] = [];
    let joinToNext = false;
    for (const inputLine of input.split('\n')) {
        const adjustedLine = adjustLine(inputLine, joinToNext);
        saveLine(outputLines, joinToNext, adjustedLine);

        // Decide what to do with the next line:
        joinToNext = endsWith1Slash(inputLine);
    }
    return outputLines.join('\n');
}

/**
 * Take an input string and split it into a list of statements.
 *
 * Generally this is similar to splitting the string into lines, but handles line
 * continuations and escape sequences.
 *
 * @param input Input string
 * @returns List of statements
 */
export function scan(input: string): string[] {
    return continue_lines(input)
        .split('\n')
        .map((rawLine: string) => rawLine.trim())
        .filter((line) => line !== '');
}
