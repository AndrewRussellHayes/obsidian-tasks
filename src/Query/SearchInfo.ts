import type { Task } from '../Task';

/**
 * SearchInfo will soon contain selected data passed in from the {@link Query} being executed.
 *
 * This is the Parameter Object pattern: it is a container for information that needs
 * to be passed down through multiple levels of code, without having to update
 * the function signatures of all the layers in between.
 */
export class SearchInfo {
    /** The list of tasks being searched.
     */
    public readonly allTasks: Readonly<Task[]>;

    public constructor(allTasks: Task[]) {
        this.allTasks = [...allTasks];
    }
}
