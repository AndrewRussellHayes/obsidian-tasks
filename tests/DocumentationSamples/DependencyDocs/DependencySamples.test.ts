/**
 * @jest-environment jsdom
 */

import moment from 'moment';

import { verifyAll } from 'approvals/lib/Providers/Jest/JestApprovals';
import { StatusConfiguration, StatusType } from '../../../src/Statuses/StatusConfiguration';
import { StatusRegistry } from '../../../src/Statuses/StatusRegistry';
import { SearchInfo } from '../../../src/Query/SearchInfo';
import type { Filter } from '../../../src/Query/Filter/Filter';
import { parseFilter } from '../../../src/Query/FilterParser';
import { SampleTasks } from '../../TestingTools/SampleTasks';
import { booleanToEmoji } from '../../TestingTools/FilterTestHelpers';

window.moment = moment;

describe('blocking and blocked filters', () => {
    beforeEach(() => {
        const nonTaskStatus = new StatusConfiguration('Q', 'Question', 'A', true, StatusType.NON_TASK);
        StatusRegistry.getInstance().add(nonTaskStatus);
    });

    afterEach(() => {
        StatusRegistry.getInstance().resetToDefaultStatuses();
    });

    it('blocking and blocked', () => {
        const tasks = SampleTasks.withWideSelectionOfDependencyScenarios();
        const instructions = [
            //
            'is blocking',
            'is blocked',
        ];
        const filters: Filter[] = instructions.map((instruction) => {
            return parseFilter(instruction)!.filter!;
        });
        const searchInfo = SearchInfo.fromAllTasks(tasks);

        verifyAll('Visualise dependency-related filters on Task, for a collection of tasks', tasks, (task) => {
            let result = task.toFileLineString() + '\n';
            filters.forEach((filter) => {
                const matches = filter!.filterFunction(task, searchInfo);
                result += `    ${filter?.instruction}
        ${booleanToEmoji(matches)}\n`;
            });
            return result;
        });
    });
});
