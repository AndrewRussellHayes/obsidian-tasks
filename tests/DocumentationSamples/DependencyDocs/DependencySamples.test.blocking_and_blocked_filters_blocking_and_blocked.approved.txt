Visualise dependency-related filters on Task, for a collection of tasks


- [ ] No dependency - TODO
    is blocking
        ❌ false
    is blocked
        ❌ false

- [x] No dependency - DONE
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 1 - TODO depends on TODO 🆔 scenario1
    is blocking
        ✅ true
    is blocked
        ❌ false

- [ ] scenario 1 - TODO depends on TODO ⛔️ scenario1
    is blocking
        ❌ false
    is blocked
        ✅ true

- [x] scenario 2 - TODO depends on DONE 🆔 scenario2
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 2 - TODO depends on DONE ⛔️ scenario2
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 3 - DONE depends on TODO 🆔 scenario3
    is blocking
        ❌ false
    is blocked
        ❌ false

- [x] scenario 3 - DONE depends on TODO ⛔️ scenario3
    is blocking
        ❌ false
    is blocked
        ❌ false

- [x] scenario 4 - DONE depends on DONE 🆔 scenario4
    is blocking
        ❌ false
    is blocked
        ❌ false

- [x] scenario 4 - DONE depends on DONE ⛔️ scenario4
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 5 - TODO depends on non-existing ID ⛔️ nosuchid
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 6 - TODO depends on self 🆔 self ⛔️ self
    is blocking
        ✅ true
    is blocked
        ✅ true

- [x] scenario 7 - task with duplicated id - this is DONE                                  - 🆔 scenario7
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 7 - task with duplicated id - this is TODO - and is blocking                - 🆔 scenario7
    is blocking
        ✅ true
    is blocked
        ❌ false

- [ ] scenario 7 - TODO depends on id that is duplicated - ensure all tasks are checked    - ⛔️ scenario7
    is blocking
        ❌ false
    is blocked
        ✅ true

- [ ] scenario 8 - mutually dependant 🆔 scenario8a ⛔️ scenario8b
    is blocking
        ✅ true
    is blocked
        ✅ true

- [ ] scenario 8 - mutually dependant 🆔 scenario8b ⛔️ scenario8a
    is blocking
        ✅ true
    is blocked
        ✅ true

- [ ] scenario 9 - cyclic dependency 🆔 scenario9a ⛔️ scenario9c
    is blocking
        ✅ true
    is blocked
        ✅ true

- [ ] scenario 9 - cyclic dependency 🆔 scenario9b ⛔️ scenario9a
    is blocking
        ✅ true
    is blocked
        ✅ true

- [ ] scenario 9 - cyclic dependency 🆔 scenario9c ⛔️ scenario9b
    is blocking
        ✅ true
    is blocked
        ✅ true

- [ ] scenario 10 - multiple dependencies TODO         - 🆔 scenario10a
    is blocking
        ✅ true
    is blocked
        ❌ false

- [/] scenario 10 - multiple dependencies IN_PROGRESS  - 🆔 scenario10b
    is blocking
        ✅ true
    is blocked
        ❌ false

- [x] scenario 10 - multiple dependencies DONE         - 🆔 scenario10c
    is blocking
        ❌ false
    is blocked
        ❌ false

- [-] scenario 10 - multiple dependencies CANCELLED    - 🆔 scenario10d
    is blocking
        ❌ false
    is blocked
        ❌ false

- [Q] scenario 10 - multiple dependencies NON_TASK     - 🆔 scenario10e
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 10 - multiple dependencies              - ⛔️ scenario10a,scenario10b,scenario10c,scenario10d,scenario10e
    is blocking
        ❌ false
    is blocked
        ✅ true

- [ ] scenario 11 - indirect dependency - indirect blocking of scenario11c ignored - 🆔 scenario11a
    is blocking
        ❌ false
    is blocked
        ❌ false

- [x] scenario 11 - indirect dependency - DONE                                     - 🆔 scenario11b ⛔️ scenario11a
    is blocking
        ❌ false
    is blocked
        ❌ false

- [ ] scenario 11 - indirect dependency - indirect blocking of scenario11a ignored - 🆔 scenario11c ⛔️ scenario11b
    is blocking
        ❌ false
    is blocked
        ❌ false

